torch/distributed/distributed_c10d.py

```diff
@@ -1465,6 +1466,126 @@ def reduce_scatter(output,
         work.wait()
 
 
+def alltoall(output,
+             input,
+             output_split_sizes=[],
+             input_split_sizes=[],
+             group=group.WORLD,
+             async_op=False):
```

Comment:
Hello @ddkalamk, really nice work! Couple of suggestions on the API and implementations.
1. Looks like there are two "flavors" of AlltoAll in this PR: a version that logically splits a single tensor into 'N' partitions on each rank and one that already takes a pre-split list of tensors. I wonder if it would make sense to also have two APIs _or_ only support the "list-of-tensors" form. That way, it's clear when the splits should be specified. One thought I had for signatures was:
   * `def alltoall_list(inputs: List[Tensor]) -> List[Tensor]` and
   * `def alltoall_single(input: Tensor, dimension: int, splits:List[int]) -> Tensor`. This will logically partition `input` into `nproc` partitions along `dimension` using `splits`. Optionally if `splits` is None or [], you can follow a partitioning strategy similar to ScaLAPACK's one-dimensional block distribution (https://www.netlib.org/scalapack/slug/node75.html#figbcol)  
2. Ideally, all output metadata should be computable just form inputs. The downside is that computing output displacements from input displacements requires an additional alltoall or allgather but typically with alltoall (at least in my experience) you do need to do this step anyway and most of the time goes in communicating the actual arrays and not the initial displacement data. Additionally, this makes using the API much less error-prone. 
3. I saw that MPI now supports non-blocking versions of alltoall/alltoallv. Have you considered supporting these as well?

Reply:
Thanks, Sudarshan!

2.> I think that's a convenient option. My guess if we orchestrate learning from a single source then we can avoid this round of communication for uneven splits as well. I presume for even splits it would be similar to other collective communication where we don't require extra round of communication to ensure sanity of the inputs.

Let's maybe offer this as an option as a followup PR if necessary? (this PR might get too complicated)

Reply:
Thanks Sudarshan and Alexander for your feedback. Here are my thoughts.
1.> The current distributed API only support `XX_list()` versions, e.g. `scatter`, `gather` or `allgather`. All of these take list of tensors and none has option to take single tensor and split internally. However, disadvantage of this is we add overhead of copying them to contiguous buffer inside MPI backend. To avoid that I have added single tensor version. I believe the newly added ProcessGroup::allgather_base() has similar motivation. However, there is no corresponding change added to python API. Therefore, I just overloaded current API for both the semantics. I think it is more of a API naming decision. If you prefer to name them as `alltoall_list` and `alltoall_single`, i will make that change.

2.> Computing output splits is one time job and typically done in the init phase. Doing it inside API would add significant performance overhead. I guess that's reason MPI_Alltoall allows it to specify explicitly.

3.> current MPI backend uses dedicated thread to drive communication. Therefore, from pytorch application point of view, communication is still asynchronous. Using async MPI collective calls would require changes to core MPI backend implementation and I think out of scope for this PR.

Reply:
Thanks, here's my two cents but I'll defer to the experts like @agolynski  and @srinivas212
1) On the API, having one flavor of alltoall that follows the convention of existing collectives and works with a list of tensors might be sufficient for the first version? Slicing tensors can be cheap if they're all just views of the same larger tensor. 
2) On computing recv offsets inside the API: This adds an overhead of a single alltoall with scalars which is usually tiny compared to the subsequent alltoall or alltoallv to exchange the actual data. On the positive side, this less error-prone for the user.  My two cents would be towards a simpler API for now and add additional knobs later if needed for performance.  But I'll defer to the experts on this.
3) Sounds good, thanks for explaining.

Reply:
1. Agree with you that with current convention of existing collectives, only version with list of tensors fits well. However, I added single tensor version looking at newly added `allgather_base` API. The discussion about using single tensor instead of list of tensor applies to almost all collectives and they all need common mechanism to support uneven splits.
2. About computing receive offsets, if we focus on list of tensor version of API, we need this information before making `alltoall` call to create list of correct sized output tensors. This is very similar to existing `gather` or `allgather` APIs which take list of output tensors. The single tensor version is just extension of this conventional list based version where we are passing size information as part of list of `split_sizes`.

Reply:
@drdarshan re. 2> I think there's another issue as well. It might be a bit tricky to implement the case you refer to because the background thread is simply executing one collective call at a time. We will need queue up two collective calls and have the background thread have the output of first one feed into the second one. On the flip side, it also does not make sense to do this from the main thread because it would block the main thread exposing the communication overheads.

Reply:
Hello @ddkalamk, thanks and you're right. Let's keep the API consistent with the one for `distributed.all_gather`. Maybe down the line, we can revise the collectives together (or provide utility functions) to compute output shapes automatically.